vulnfeeds/cvelist2osv/grouping.go

```diff
@@ -0,0 +1,174 @@
+package cvelist2osv
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/ossf/osv-schema/bindings/go/osvschema"
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/types/known/structpb"
+)
+
+// groupAffectedRanges groups ranges that share the same introduced value, type, and repo.
+// This is because having multiple ranges with the same introduced value would act like an
+// OR condition, rather than AND.
+func groupAffectedRanges(affected []*osvschema.Affected) {
+	for _, aff := range affected {
+		if len(aff.GetRanges()) <= 1 {
+			continue
+		}
+
+		// Key for grouping: Type + Repo + Introduced Value
+		type groupKey struct {
+			RangeType  osvschema.Range_Type
+			Repo       string
+			Introduced string
+		}
+
+		groups := make(map[groupKey]*osvschema.Range)
+		var order []groupKey // To maintain deterministic order of first appearance
+
+		for _, r := range aff.GetRanges() {
+			// Find the introduced event
+			var introduced string
+			for _, e := range r.GetEvents() {
+				if e.GetIntroduced() != "" {
+					introduced = e.GetIntroduced()
+					break
+				}
+			}
+
+			// If no introduced event is found, we use an empty string as the introduced value.
+			key := groupKey{
+				RangeType:  r.GetType(),
+				Repo:       r.GetRepo(),
+				Introduced: introduced,
+			}
+
+			if _, exists := groups[key]; !exists {
+				// Initialize with a deep copy of the first range found for this group
+				// We need to be careful about DatabaseSpecific.
+				// We want to keep the "versions" from this first range.
+				groups[key] = &osvschema.Range{
+					Type:             r.GetType(),
+					Repo:             r.GetRepo(),
+					Events:           []*osvschema.Event{},
+					DatabaseSpecific: r.GetDatabaseSpecific(), // Start with this one's DS
+				}
+				order = append(order, key)
+			} else {
+				// Merge DatabaseSpecific "versions"
+				mergeDatabaseSpecificVersions(groups[key], r.GetDatabaseSpecific())
+			}
+
+			// Add all events to the group. Deduplication happens later in cleanEvents.
+			groups[key].Events = append(groups[key].Events, r.GetEvents()...)
+		}
+
+		// Reconstruct ranges from groups
+		var newRanges []*osvschema.Range
+		for _, key := range order {
+			r := groups[key]
+			r.Events = cleanEvents(r.GetEvents())
+			newRanges = append(newRanges, r)
+		}
+		aff.Ranges = newRanges
+	}
+}
+
+func mergeDatabaseSpecificVersions(target *osvschema.Range, source *structpb.Struct) {
+	if source == nil {
+		return
+	}
+	sourceVersions := source.GetFields()["versions"]
+	if sourceVersions == nil {
+		return
+	}
+
+	if target.GetDatabaseSpecific() == nil {
+		target.DatabaseSpecific = &structpb.Struct{Fields: make(map[string]*structpb.Value)}
```

Comment:
nit: I feel [`structpb.NewStruct(nil)`](https://pkg.go.dev/google.golang.org/protobuf/types/known/structpb#NewStruct) may be more idiomatic (though I'm unsure if `nil` is okay to use in that function)